### PR TITLE
Added widget template with unit tests

### LINKED ISSUE #103: Add unit tests to templates
To demonstrate how to test extension, especially using DI.

### PR DESCRIPTION
**Pull request by JonasHelming**
fixed #103

Adds an extended version of the widget template including unit tests.

How to test:
- Create the template "Widget with Unit Tests"
- Run tests with "yarn test"

Contributed on behalf of STMicroelectronics

Signed-off-by: Jonas Helming <jhelming@eclipsesource.com>

### CODE REVIEW

**Review by vince-fugnitto — COMMENTED** [review]
@JonasHelming I'd be fine with adding the unit-tests directly in the `widget` template rather than splitting it up into two templates. Any reason we want to do the split?

**Comment by JonasHelming** [comment]
Good question! Only reason would be separation of concerns and keep the plain widget template simpler. So you would prefer to merge them?

**Comment by JonasHelming** [comment]
Please note, that I will contribute a template for Ui tests as well

**Comment by vince-fugnitto** [comment]
@JonasHelming I think merging them where we have the `widget` extension with `unit-tests` already available would be fine in my opinion. I would not use `jest` however as it is quite heavy, perhaps we should use what we have in the main repo (mocha).

**Comment by vince-fugnitto** [comment]
I don't have a strong opinion of using mocha (or another testing framework) vs jest, so I'll leave the decision up to you :)

**Comment by JonasHelming** [comment]
merged!

File: src/app/index.ts [hunk 1]
```diff
@@ -148,7 +152,7 @@ module.exports = class TheiaExtension extends Base {
                 message: "The extension's type",
                 choices: [
                     { value: ExtensionType.HelloWorld, name: 'Hello World' },
-                    { value: ExtensionType.Widget, name: 'Widget' },
+                    { value: ExtensionType.Widget, name: 'Widget (with unit tests)' },
```
Comment (vince-fugnitto) [inline src/app/index.ts:155]: I'd drop the 'with unit tests' in the name:

```suggestion
                    { value: ExtensionType.Widget, name: 'Widget' },
```

We should update the `readme` of the generated extension to include mentions of included unit-tests and how to run them (which is already done).

File: src/app/index.ts [hunk 2]
```diff
@@ -320,7 +329,22 @@ module.exports = class TheiaExtension extends Base {
                 this.extensionPath('README.md'),
                 { params: this.params }
             );
-        }
+            this.fs.copyTpl(
+                this.templatePath('widget/__tests__/widget.test.ts'),
```
Comment (vince-fugnitto) [inline src/app/index.ts:333]: We should follow the format of other tests found in the upstream repo (drop `__tests__` as a name).

File: src/app/index.ts [hunk 3]
```diff
@@ -320,7 +329,22 @@ module.exports = class TheiaExtension extends Base {
...
+                this.extensionPath(`src/browser/__tests__/${this.params.extensionPath}-widget.test.ts`),
+                { params: this.params }
+            );
+            this.fs.copyTpl(
+                this.templatePath('widget/__tests__/mock-objects/mock-message-service.ts'),
+                this.extensionPath(`src/browser/__tests__/mock-objects/mock-message-service.ts`),
+                { params: this.params }
+            );
```
Comment (vince-fugnitto) [inline src/app/index.ts:341]: It might be overkill to completely mock the `message-service`.

File: templates/widget/__tests__/widget.test.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import 'reflect-metadata';
+import { MessageClient, MessageService } from '@theia/core';
+import { ContainerModule, Container } from 'inversify';
+import { MockMessageClient, MockMessageService } from './mock-objects/mock-message-service';
+import { <%= params.extensionPrefix %>Widget } from '../<%= params.extensionPath %>-widget';
+import { render } from '@testing-library/react'
+
+describe('widget extension unit tests', () => {
+
+    let widget: any;
```
Comment (vince-fugnitto) [inline templates/widget/__tests__/widget.test.ts:10]: We should include the actual typings else we have no autocompletion and it leads to odd code later on.

File: templates/widget/__tests__/widget.test.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    beforeEach(async () => {
+        const module = new ContainerModule( bind => {
+            bind(MessageClient).to(MockMessageClient).inSingletonScope();
+            bind(MessageService).to(MockMessageService).inSingletonScope();
```
Comment (vince-fugnitto) [inline templates/widget/__tests__/widget.test.ts:15]: We can mock what we need:

```suggestion
            bind(MessageService).toConstantValue({
                info(message: string): void {
                    console.log(message);
                }
            } as MessageService);
```

File: templates/widget/__tests__/widget.test.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import 'reflect-metadata';
+import { MessageClient, MessageService } from '@theia/core';
+import { ContainerModule, Container } from 'inversify';
+import { MockMessageClient, MockMessageService } from './mock-objects/mock-message-service';
+import { <%= params.extensionPrefix %>Widget } from '../<%= params.extensionPath %>-widget';
+import { render } from '@testing-library/react'
+
+describe('widget extension unit tests', () => {
```
Comment (vince-fugnitto) [inline templates/widget/__tests__/widget.test.ts:8]: We should include actual name of the widget:

```suggestion
describe('<%= params.extensionPrefix %>Widget', () => {
```

File: templates/widget/__tests__/widget.test.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import 'reflect-metadata';
```
Comment (vince-fugnitto) [inline templates/widget/__tests__/widget.test.ts:1]: The name of the file should likely align with the name of the actual widget.

File: templates/widget/__tests__/widget.test.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    it('should render react node correctly', async () => {
+        const element = render(widget.render());
+        expect(element.queryByText('Display Message')).toBeTruthy();
+    });
+
+    it('should inject the message service', async () => {
+        widget.displayMessage();
+        expect(widget.messageService.infoWasCalled).toBe(true);
+    });
```
Comment (vince-fugnitto) [inline templates/widget/__tests__/widget.test.ts:31]: This test is likely overkill for what it actually tests.

We can do it the following way (without the need to mock the service):

```ts
describe('widget extension unit tests', () => {

    let widget: any;

    beforeEach(async () => {
        const module = new ContainerModule(bind => {
            bind(MessageClient).to(MockMessageClient).inSingletonScope();
            bind(MessageService).toConstantValue({
                info(message: string): void {
                    console.log(message);
                }
            } as MessageService);
            bind(CustomWidgetWidget).toSelf();
        });
        const container = new Container();
        container.load(module);
        widget = container.resolve<CustomWidgetWidget>(CustomWidgetWidget);
    });

    it('should be an instance of \'CustomWidgetWidget\'', () => {
        expect(widget instanceof CustomWidgetWidget).toBeTruthy();
    });

    it('should render react node correctly', async () => {
        const element = render(widget.render());
        expect(element.queryByText('Display Message')).toBeTruthy();
    });

    it('should inject \'MessageService\'', () => {
        const spy = jest.spyOn(widget, 'displayMessage')
        widget.displayMessage();
        expect(spy).toBeCalled();
    });

});
```

File: templates/widget/__tests__/widget.test.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+});
```
Comment (vince-fugnitto) [inline templates/widget/__tests__/widget.test.ts:33]: newline.

File: templates/widget/configs/jest.config.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import type { Config } from '@jest/types';
+
+export default async (): Promise<Config.InitialOptions> => ({
+    preset: 'ts-jest',
+    testMatch: ['**.test.ts'],
+    rootDir: '../',
+    transform: {
+        '^.+\\.(ts)$': 'ts-jest',
+    }
+});
```
Comment (vince-fugnitto) [inline templates/widget/configs/jest.config.ts:10]: newline.

File: src/app/index.ts (hunk 1 above)
Comment (JonasHelming) [reply src/app/index.ts:155]: I do not have a strong opinion on this, but keeping (with unit tests) allows users to find an example for unit tests, so I prefer to keep it.

File: templates/widget/__tests__/widget.test.ts (hunk 7 above)
Comment (JonasHelming) [reply templates/widget/__tests__/widget.test.ts:1]: I am not sure what this means? The file name is parameterized.

File: templates/widget/__tests__/widget.test.ts (hunk 8 above)
Comment (JonasHelming) [reply templates/widget/__tests__/widget.test.ts:31]: Yes, we were discussing the trade off between adding a "spyOn" as an additional concept an potential complexity vs. a dedicated Mock. However, looking at your proposal, I am convinced that this is more comprehensible. I adapted it.

File: templates/widget/__tests__/widget.test.ts (hunk 4 above)
Comment (JonasHelming) [reply templates/widget/__tests__/widget.test.ts:10]: Yes, did that, but had to adapt the signature of widget.

File: src/app/index.ts (hunk 1 above)
Comment (JonasHelming) [reply src/app/index.ts:155]: I would leave it and remove it when other templates also have tests

Comment (vince-fugnitto) [reply src/app/index.ts:155]: I'll leave the decision up to you, I think that is a fine proposal to only drop it when other templates also include tests.

File: templates/widget/__tests__/widget.test.ts (hunk 7 above)
Comment (vince-fugnitto) [reply templates/widget/__tests__/widget.test.ts]: I see now that the file path is parameterized, just that this is the template path :+1: 
https://github.com/eclipse-theia/generator-theia-extension/pull/112/files#diff-e234283aa9cd1cdf0803a13194964671ae1dfd31064f5fe35429df36a51ef17bR334

File: templates/widget/__tests__/widget.test.ts (hunk 4 above)
Comment (vince-fugnitto) [reply templates/widget/__tests__/widget.test.ts]: @JonasHelming please note that the following can remain `protected` and we can instead update the test:

```ts
protected displayMessage(): void {
    this.messageService.info('Congratulations: FooBar Widget Successfully Created!');
}
```

```ts
it('should inject \'MessageService\'', () => {
    const spy = jest.spyOn(widget as any, 'displayMessage')
    widget['displayMessage']();
    expect(spy).toBeCalled();
});
```

I'll leave the decision up to you, but I feel as though the `displayMessage` should remain `protected` as it should not allow outside callers when developers are modifying the template.

Comment (JonasHelming) [reply templates/widget/__tests__/widget.test.ts]: Yeah, was torn between two options, now took yours...

File: templates/widget/widget.test.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import 'reflect-metadata';
+import { MessageService } from '@theia/core';
+import { ContainerModule, Container } from 'inversify';
+import { <%= params.extensionPrefix %>Widget } from './<%= params.extensionPath %>-widget';
+import { render } from '@testing-library/react'
+
+describe('<%= params.extensionPrefix %> widget extension unit tests', () => {
```
Comment (vince-fugnitto) [inline templates/widget/widget.test.ts:7]: I believe it should only be the name of the widget?

```suggestion
describe('<%= params.extensionPrefix %>Widget', () => {
```

**Review by vince-fugnitto — COMMENTED** [review]
My last minor comment is the name within the `describe('')`.

**Review by vince-fugnitto — APPROVED** [review]
@JonasHelming the changes look good to me, can you please use the `squash and merge` option when merging so we can have a clean history?

**Comment by JonasHelming** [comment]
Sure, thank you very much for the review, Vince!

**Pull request merged**